### ISSUE TITLE
Implement ANALYZE on target variables

### ISSUE THREAD

**Issue by fsaad** [probcomp/bayeslite#429]
### Problem

For a bayeslite generator `G`, MML supports running a generic "transition" operator:

`ANALYZE G FOR <s> SECONDS WAIT`

We are now at a stage where this per-generator `ANALYZE` is too coarse-grained. In a high-dimensional population `pop`, a user may only want to analyze a subset of the modeled variables. For a composite GPM with sub-cgpms consisting of `gpmcc` and others i.e. `linear_regression`, the runtime characteristics of `ANALYZE` will vary wildly -- `gpmcc` might need several hours, while `linear_regression` just needs a few minutes -- and it does not make sense to intertwine the transition operators in one big `ANALYZE` statement. Here is a working example:

```
CREATE POPULATION pop FROM pop.csv(
    a NUMERICAL,
    b NUMERICAL,
    ... 
    y NUMERICAL,
    z NUMERICAL
)

CREATE GENERATOR G FOR pop USING cgpm(
    a LOGNORMAL,
    b EXPONENTIAL,
    -- Default components model for c through z.

    -- Override gpmcc for variable y.
    MODEL y GIVEN x, z USING linear_regression;
)
```

Currently, the `cgpm_metamodel` exercises an enormous hack [over here](https://github.com/probcomp/bayeslite/blob/20160624-riastradh-cgpm/src/metamodels/cgpm_metamodel.py#L220-L224) where the hooked cgpms are transition for 1 `iteration` at composition time, and then never again!
### Solution

Extend `ANALYZE` to support taking in an optional set of target variables. As it stands `gpmcc` accepts a set of target variables to transition against. For a composite cgpm, the runtime will look up the cgpms that model those variables, and use `cgpm.transition(N, targets)` . All this logic will be implemented by `Engine` and thus transparent to `bayeslite` (for model independence).

Here are some possible ways we can extend `ANALYZE` to gracefully deal with this situation:
#### Option 1

```
ANALYZE G
    TARGETING <vars, ...>
    FOR <n> ITERATION

ANALYZE G
    TARGETING *
    FOR <n> ITERATION
```
#### Option 2

```
ANALYZE <vars, ...>
    FROM G
    FOR <n> ITERATION;

ANALYZE *
    FROM G
    FOR <n> ITERATION;
```
#### Options 3+

Please weigh in!

The question is whether `*` should apply to all variables, or only those modeled by `gpmcc`. It could be a poor UI to have the user specify 25 variables modeled by `gpmcc`, leaving only 1 out (i.e. `a` through `z` except for `y` in the example above) if they do not want to transition the composed cgpms. One option could be

```
ANALYZE *
    FROM G
    [IGNORE | EXCEPT FOR | SKIP | etc] <vars, ...>
    FOR <n> ITERATION;
```

Another issue is coordinating the time for analysis for variables that come from different CGPMs, i.e. when a user says `FOR <s> SECONDS`. One solution is to disallow timed analysis when the target variables that are modeled by different cgpms. Timed analysis in `lovecat` is sub-optimal anyway since it is not parallelized at all, but `gpmcc` can do multiprocessed timed analysis.

**Comment by riastradh-probcomp** [comment]
First reaction: I anticipate that we will want to develop a richer and possibly per-metamodel language for programming analysis, much like we already have for expressing generators.  For example, it is not hard to imagine that we will want to be able to embed VentureScript for some things.

    ANALYZE satellites_lovcat FOR 360 MINUTES (
        VARIABLES (x, y, z),
        HYPERPARAMETERS,
        ...
    );
    ANALYZE satellites_cgpm FOR 360 MINUTES (
        VARIABLES (x, y, z),
        VARIABLES (apogee, perigee, period)
            VENTURESCRIPT { mh(default, one, 25) },
        ...
    );

On `*`: I can't imagine why you would want to make `*` apply to all variables including those that aren't modelled and hence can't be analyzed, which would render it unusable for `ANALYZE`.  It seems intuitively obvious to me that if there is any `*` notation it should apply only to modelled variables -- all variables that could possibly make sense in the context.

Timed analysis: Obviously we should have a serious story for timed analysis that is not lock-step deparallelized like lovecat is, by pushing the timing into the metamodel code.  The only reason we did not do that for lovecat was vkm's consistent exhortations to touch lovecat as little as possible.

**Comment by fsaad** [comment]
@riastradh-probcomp yes, we talked about "more advanced" analysis languages for metamodels. I am bound to push that for the future, until we have a more concrete proposal for the MML syntaxes and use cases of custom inference programs.

I was reading the BayesDB paper, and the phrase that appears there for `ANALYZE` is

```
ANALYZE [variable subset OF] population FOR timelimit
```

In summary, I agree with you, but suggest we first do the variable subset, which we understand, is related to our immediate/medium goals, and we can hit the ground running directly using cgpm-lib (which takes in target variables, and I am adding the ability for State to transition hooked cgpms as well).

**Comment by riastradh-probcomp** [comment]
The reason I suggest the foundation of a more generic language is not that I think we should go beyond a simple set of target variables now, but rather that the work done to get

    ANALYZE population FOR duration (
        VARIABLES (x, y, z)
    )

working and factored generically enough for future use (a) is not terribly difficult and (b) would not be wasted for, or get in the way of, future extensions.

**Comment by fsaad** [comment]
Related https://github.com/probcomp/bayeslite/issues/571

**Comment by fsaad** [comment]
Fixed by e4138e1 which resolved  https://github.com/probcomp/bayeslite/issues/571
